source/credentials_provider_ecs.c

```diff
@@ -0,0 +1,636 @@
+/*
+ * In general, the ECS document looks something like:
+
+{
+  "Code" : "Success",
+  "LastUpdated" : "2019-05-28T18:03:09Z",
+  "Type" : "AWS-HMAC",
+  "AccessKeyId" : "...",
+  "SecretAccessKey" : "...",
+  "Token" : "...",
+  "Expiration" : "2019-05-29T00:21:43Z"
+}
+
+ */
+static struct aws_credentials *s_parse_credentials_from_ecs_document(
+    struct aws_allocator *allocator,
+    struct aws_byte_buf *document) {
+
+    struct aws_credentials *credentials = NULL;
+    cJSON *document_root = NULL;
+
+    struct aws_byte_cursor null_terminator_cursor = aws_byte_cursor_from_string(s_empty_empty_string);
+    if (aws_byte_buf_append_dynamic(document, &null_terminator_cursor)) {
+        goto done;
+    }
+
+    document_root = cJSON_Parse((const char *)document->buffer);
+    if (document_root == NULL) {
+        goto done;
+    }
+
+    /*
+     * Pull out the three credentials components
+     */
+    cJSON *access_key_id = cJSON_GetObjectItemCaseSensitive(document_root, aws_string_c_str(s_access_key_id_name));
+    if (!cJSON_IsString(access_key_id) || (access_key_id->valuestring == NULL)) {
+        goto done;
+    }
+
+    cJSON *secret_access_key =
+        cJSON_GetObjectItemCaseSensitive(document_root, aws_string_c_str(s_secret_access_key_name));
+    if (!cJSON_IsString(secret_access_key) || (secret_access_key->valuestring == NULL)) {
+        goto done;
+    }
+
+    cJSON *session_token = cJSON_GetObjectItemCaseSensitive(document_root, aws_string_c_str(s_session_token_name));
+    if (!cJSON_IsString(session_token) || (session_token->valuestring == NULL)) {
+        goto done;
+    }
+
+    cJSON *creds_expiration =
+        cJSON_GetObjectItemCaseSensitive(document_root, aws_string_c_str(s_creds_expiration_name));
+    if (!cJSON_IsString(creds_expiration) || (creds_expiration->valuestring == NULL)) {
+        goto done;
+    }
+
+    /*
+     * Build the credentials
+     */
+    struct aws_byte_cursor access_key_id_cursor = aws_byte_cursor_from_c_str(access_key_id->valuestring);
+    struct aws_byte_cursor secret_access_key_cursor = aws_byte_cursor_from_c_str(secret_access_key->valuestring);
+    struct aws_byte_cursor session_token_cursor = aws_byte_cursor_from_c_str(session_token->valuestring);
+    struct aws_byte_cursor creds_expiration_cursor = aws_byte_cursor_from_c_str(creds_expiration->valuestring);
+
+    if (access_key_id_cursor.len == 0 || secret_access_key_cursor.len == 0 || session_token_cursor.len == 0) {
+        goto done;
+    }
+
+    credentials = aws_credentials_new_from_cursors(
+        allocator, &access_key_id_cursor, &secret_access_key_cursor, &session_token_cursor);
+
+    if (creds_expiration_cursor.len != 0) {
+        struct aws_date_time expiration;
+        aws_date_time_init_from_str_cursor(&expiration, &creds_expiration_cursor, AWS_DATE_FORMAT_ISO_8601);
+        credentials->expiration_timepoint_seconds = expiration.timestamp;
```

Comment:
don't access the internal members of `aws_data_time` directly, use the public API. I think you want `aws_date_time_as_epoch_secs()`?
If API functions exist for a struct, you should consider its member variables private. Yes, we declared `struct aws_date_time` in a header, but only did that so it can be used without requiring a heap allocation.

The only exceptions where it's ok to access the member variables are:
- `aws_byte_cursor`
- "options" structs like `aws_credentials_provider_ecs_options` that have no API functions.
- an implementation can touch its own members of course